Add helper to update workflows using Github or a remote server

spamwax opened the pull request:
- Adds a new module to let workflow authors update their
workflow's bundle using either github.com or some other remote service.
- Adds documentation for the new helper module.
- Add tests and update .travis.yml
- Use `cargo doc --features updater` to generate
docs.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/kballard/alfred-rs/8)
<!-- Reviewable:end -->

spamwax committed:
Add helper to update workflows using Github or a remote server

- Adds a new module to let workflow authors update their
workflow's bundle using either github.com or some other remote service.
- Adds documentation for the new helper module.
- Add tests and update .travis.yml
- Use `cargo doc --features updater` to generate
docs.

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+//! ```
+//! Using this module, the workflow author will be able to make Alfred
+//! check for & download latest releases from the remote server
+//! within ajustable intervals (default is 24 hrs).
````
Typo: "ajustable" should be "adjustable"

lilyball commented on src/lib.rs:
```diff
@@ -7,6 +7,13 @@
+//! Additionally the crate provides a way of checking for and updating
+//! your workflows automatically (Alfred 3 and above only).
+//!
+//! See [`updater`] module documentation for details and examples.
```
I checked, this actually hyperlinks in `cargo doc`, but I'm confused because this isn't valid Markdown. In Markdown it would have to be

    [`updater`][]

Do you know why this works without the trailing `[]`?

spamwax replied:
I think this is working because in next line I have defined what `[updater]` is pointing to.
However I ran to similar issues when using nightly/stable. It seems rustdoc is moving to a new markdown backend and this is related to the transition.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//!
+//! For convenience, an associated method [`Updater::gh()`] is available to check for workflows hosted on `github.com`.
+//!
+//! However, it's possible to check with other servers as long as [`Releaser`] trait is
```
"as long as Releaser trait" should probably say "as long as the Releaser trait"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! attach/upload `YourWorkflow.alfredworkflow` to the release page.
+//!
+//! The tag should follow all the [semantic versioning] rules.
+//! The only exception to those rulse is that you can prepend your
```
Typo: "rulse" should be "rules"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! Alfred in its preferences window.
+//!
+//! ### Note to workflow authors
+//! - Workflolw authors should make sure that _released_ workflow files have
```
Typo: "Workflolw" should be "Workflow"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//!
+//! // Above will save the state of `Updater` in workflow's data folder.
+//! // Depending on how long has elapsed since first run consequent calls
+//! // to `update_ready()` may retuan false if it has been less than
```
Typo: "retuan" should be "return"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! // Above will save the state of `Updater` in workflow's data folder.
+//! // Depending on how long has elapsed since first run consequent calls
+//! // to `update_ready()` may retuan false if it has been less than
+//! // intervals set for checking (defaults to 24 hours).
```
"intervals" should probably be "the interval"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! // intervals set for checking (defaults to 24 hours).
+//!
+//! // However in subsequent runs, when the checking interval period has elapsed
+//! // and there actually exist a new release, then `update_ready()` will return true.
```
"exist" should probably be "exists"

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+//! ```
+//!
+//! For the above example to automatically work, you then need to connect the output of the script
+//! to an **Open File** action so that Alfred can insall/upgrade the new version.
````
Typo: "insall" should be "install"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! For the above example to automatically work, you then need to connect the output of the script
+//! to an **Open File** action so that Alfred can insall/upgrade the new version.
+//!
+//! As suggested in above exmaple, you can add an Alfred variable to the item so that your workflow
```
Typo: "exmaple" should be "example"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+}
+
+impl Updater<GithubReleaser> {
+    /// Create an `Updater` object that will talk to a `github` repository to download latest releases.
```
You should probably put "the" before "latest releases".

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+impl Updater<GithubReleaser> {
+    /// Create an `Updater` object that will talk to a `github` repository to download latest releases.
+    ///
+    /// the `repo_name` should be in `user_name/repository_name` form.
```
The beginning of this sentence should be capitalized.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// # }
+    ///
+    /// # fn main() {
+    /// #     ex_new();
```
This example is going to actually try to run the updater against the bogus `user_name/repo_name`. Surely we don't want to actually run this in the tests, but instead just compile? The only thing running it does is ensure it doesn't panic.

spamwax replied:
Not sure what the fix would be here. Although the `Updater` is only initiated and no network call is made here, I think it initiates it with default values without even reading from disk.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+//! # }
+//! #
+//! # fn main() {
+//! #     if let Err(_) = run() {
```
We probably don't want to actually run this during the tests, because it's going to actually hit up your repo to look for a release and potentially download it. This is not useful behavior to do during a test (in fact, a unit test really shouldn't be hitting the network at all). So we should instead just make sure this compiles.

spamwax replied:
so should I remove the `main` function then?

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+    /// # }
+    /// ```
+    ///
+    /// `Updater` created using this method will look at the assets
````
You should probably insert "The" before "Updater"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// ends in `alfred3workflow` or `alfredworkflow`.
+    pub fn gh<S>(repo_name: S) -> Self
+    where
+        S: Into<String> + AsRef<str>,
```
Either the `Into<String>` or the `AsRef<str>` is unnecessary (the way this function is written it's the `Into<String>` that's unnecessary, but if you remove the call to `.as_ref()` then the `AsRef<str>` is the one that becomes unnecessary).

I'd suggest keeping `Into<String>` and dropping `AsRef<str>`, since you're passing the argument to something that itself wants `Into<String>`.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+where
+    T: Releaser,
+{
+    /// Create an `Updater` that will check & download latest releases from a remote server.
```
"latest releases" should probably be "the latest release".

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+{
+    /// Create an `Updater` that will check & download latest releases from a remote server.
+    ///
+    /// How the `Updater` interacts with remote server should be implemented using the [`Releaser`]
```
"remote server" should probably be "the remote server"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    ///         RemoteCIReleaser {}
+    ///     }
+    ///     fn downloadable_url(&self) -> Result<String, io::Error> {
+    ///         Ok("ci.remote.cc".to_string())
```
This one is confusing, the method is `downloadable_url` but you're returning something that doesn't appear to be a URL.

spamwax replied:
I was thinking of using the `url` crate for this. However, I need to have the `Result` there as the `Releaser` may error out trying to fetch the url from remote server.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// use alfred::Updater;
+    /// use alfred::updater::Releaser;
+    /// # use std::env;
+    /// # fn ex_remote_releaser() {
```
BTW this could just be `# fn main() {`.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// - The workflow version cannot be parsed as semver compatible identifier.
+    pub fn new<S>(repo_name: S) -> Updater<T>
+    where
+        S: Into<String> + AsRef<str>,
```
As before, one of `Into<String>` or `AsRef<str>` is unnecessary.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// # Panics
+    /// The method will panic if:
+    /// - the passed value `version` cannot be parsed as a semver compatible string the
+    /// function will panic.
```
Remove "the function will panic", it's redundant.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// # extern crate alfred;
+    /// # use alfred::Updater;
+    /// # use std::env;
+    /// # fn ex_set_interval() {
```
BTW this could just be `# fn main() {`.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// Method will panic if any file io error happens during saving the `Updater` data to disk.
+    pub fn set_interval(&mut self, tick: i64) {
+        self.set_update_interval(tick);
+        self.save().expect("cannot save updater data to file.");
```
Why do we persist the interval to disk? Why wouldn't we simply set it in code on every execution?

spamwax replied:
I included this so workflow authors can let workflow users change the interval.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+        self.state.current_version =
+            Version::parse(version.as_ref()).expect("version should follow semver rules.");
+        StdEnv::set_var("alfred_workflow_version", version.as_ref());
+        self.save().expect("cannot save updater data to file.");
```
Why do we persist the version to disk? Why wouldn't we simply set it in code on every execution?

One potential issue with persisting the version to disk is if my workflow initially sets the version in code, then I decide to set it in preferences instead and remove the code, anyone who already ran my workflow will have their version frozen forever at an incorrect value.

spamwax replied:
I will look into this, we can also remove the set_version all together to force everyone use the preferences.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+
+    /// Checks if a new update is availablle.
+    ///
+    /// This method will fetch the latess release information from repoository and compare it to
```
Typo: "latess" should be "latest", also "repoository" should be "repository"

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// Checks if a new update is availablle.
+    ///
+    /// This method will fetch the latess release information from repoository and compare it to
+    /// current release of the workflow. The repository should tag each release according to
```
"current release" should have "the" before it ("the current release").

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+        // check after UPDATE_INTERVAL seconds.
+        let epoch = EPOCH_TIME.parse::<DateTime<Utc>>().unwrap();
+
+        if self.last_check() == epoch {
```
Wouldn't it be simpler to have the last check time be an `Option` than to be using the epoch like this?

spamwax replied:
probably, will look into this

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+#[derive(Debug, Serialize, Deserialize)]
+struct UpdaterState {
+    current_version: Version,
+    last_check: RefCell<DateTime<Utc>>,
```
You can replace this `RefCell` with a `Cell`.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// # use std::env;
+    /// # use std::thread;
+    /// # use std::time;
+    /// # fn ex_due_to_check() {
```
This could just be `# fn main() {`.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    /// // `update_ready()`:
+    /// # updater.update_ready();
+    /// # updater.set_interval(1);
+    /// # thread::sleep(time::Duration::from_millis(1001));
```
Can you not just set the interval to `0`? Or, heck, set it to `-1`?

spamwax replied:
👍

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+            > Duration::seconds(self.update_interval())
+    }
+
+    /// Function to download and save the latest release and save into workflow's cache dir.
```
The second "and save" is redundant.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+
+    /// Function to download and save the latest release and save into workflow's cache dir.
+    ///
+    /// If download and save operations are both successful, it returns name of file in which the
```
The word "download" should have "the" before it ("If the download and save operations").

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+    ///
+    /// Within shell, it can be installed by issuing something like:
+    /// ```bash
+    /// open -a "Alfred 3" latest_release_WORKFLOW-UID.alfredworkflow
````
Probably better to use the `-b` flag instead of the `-a` flag, in case the user renames the app. This would be

    open -b com.runningwithcrayons.Alfred-3 latest_release_WORKFLOW-UID.alfredworkflow

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+    /// Or you can add "Run script" object to your workflow and use environment variables set by
+    /// Alfred to automaticallly open the downloaded release:
+    /// ```bash
+    /// open "$alfred_workflow_cache/latest_release_$alfred_workflow_uid.alfredworkflow"
````
Do you not want the `-a`/`-b` flag here?

spamwax replied:
probably, was just showing two ways of doing the same thing.

lilyball commented on src/updater/mod.rs:
````diff
@@ -0,0 +1,693 @@
+    /// open "$alfred_workflow_cache/latest_release_$alfred_workflow_uid.alfredworkflow"
+    /// ```
+    /// # Errors
+    /// Downloading latest workflow can fail if network error, file error or Alfred environment variable
````
You should insert "the" before "latest workflow" ("Downloading the latest workflow can fail")

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+    }
+}
+
+use self::releaser::str_to_io_err;
```
Why is this `use` way down here?

spamwax replied:
my way of reminding myself it's a private fn

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+                    Error::new(ErrorKind::Other, e)
+                })
+            })
+            .expect("couldn't read updater's saved data")
```
If this saved data is corrupted, this will cause the workflow to crash on launch forever (since presumably you create the `Updater` every time, as you don't know whether the interval has passed without it).

spamwax replied:
ok

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+                update_interval: UPDATE_INTERVAL,
+            };
+            let updater = Updater { state, releaser: r };
+            updater.save().expect("cannot save updater data to file.");
```
If this fails due to a non-transient error, this will cause the workflow to crash on launch forever (as it will attempt to save again every single time you run it).

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+
+        // Now creating another one, will load the updater from file
+        assert!(updater_state_fn.exists());
+        let _updater: Updater<GithubReleaser> = Updater::new("spamwax/alfred-pinboard-rs");
```
This isn't actually checking that it's loading the last updater state, it's merely checking that creating a second `Updater` doesn't panic.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+        assert!(updater.due_to_check());
+        // update should be ready since alfred-pinboard-rs
+        // already has newer than VERSION_TEST.
+        assert!(updater.update_ready().expect("couldn't check for update"));
```
I'm not thrilled with the idea that this actually talks to GitHub for real and gets the status of spamwax/alfred-pinboard-rs. Unit tests shouldn't talk to the network. Can we not mock out the network call?

spamwax replied:
we could, this was here to test actual errors that talking to a real server may cause. As you can see it's intended to be run locally. I can add an `#[ignore]` to it.

lilyball commented on src/updater/mod.rs:
```diff
@@ -0,0 +1,693 @@
+        // Force current version to be really old.
+        updater.set_version("0.0.1");
+        assert!(updater.update_ready().expect("couldn't check for update"));
+        assert!(updater.download_latest().is_ok());
```
Same here, I don't like that this is actually talking to the network.

As it is, shouldn't this test also be marked `#[cfg(not(feature = "ci"))]` since your other tests that perform networking are?

spamwax replied:
this is mocked (at the beginning of test function), it doesn't talk to network.

lilyball commented on src/updater/releaser.rs:
```diff
@@ -0,0 +1,215 @@
+#[cfg(test)]
+static MOCKITO_URL: &'static str = mockito::SERVER_URL;
+#[cfg(test)]
+pub const MOCK_RELEASER_REPO_NAME: &str = "Mock/Releaser";
```
This should probably be something that's not actually a potentially-valid repo.

lilyball commented on src/updater/releaser.rs:
```diff
@@ -0,0 +1,215 @@
+#[cfg(test)]
+pub const MOCK_RELEASER_REPO_NAME: &str = "Mock/Releaser";
+
+/// An interface for checking with remote servers to indentify latest release for an
```
Typo: "indentify" should be "identify", and it should have the word "the" after it ("to identify the latest release")

lilyball commented on src/updater/releaser.rs:
```diff
@@ -0,0 +1,215 @@
+/// An interface for checking with remote servers to indentify latest release for an
+/// Alfred workflow.
+///
+/// This trait has been implemented for `GithubReleaser` to check for newer version of a workflow
```
"newer version" should have "a" inserted before it ("to check for a newer version").

lilyball commented on src/updater/releaser.rs:
```diff
@@ -0,0 +1,215 @@
+    fn newer_than(&self, version: &Version) -> Result<bool, io::Error>;
+}
+
+// Struct to handle checking and downloaading release files from `github.com`
```
Typo: "downloaading" should be "downloading"

lilyball commented on src/updater/releaser.rs:
```diff
@@ -0,0 +1,215 @@
+
+        assert!(
+            releaser
+                .newer_than(&Version::from((0, 0, 0)))
```
As with the other tests, I'm uncomfortable having this do actual network access. Can you mock out the network call?

lilyball reviewed (changes requested):
Hi spamwax,

Thanks for this PR! This functionality looks awesome. I've done a full review, most of my comments are grammatical/typographical errors in the doc comments, but there are some code comments as well.

That said, I'm not sure this is really the right approach to take. What this does is once every 24 hours (by default), running the workflow will potentially block while it talks to GitHub over the network, then attempts to download the file, then finally returns. This has several problems:

1. Blocking in general is a poor experience. Especially if the user has a bad internet connection and the release download is large.
2. If the workflow is e.g. a script filter and is configured to cancel the existing run whenever you type a new character (as my workflows are), this really won't work at all, because the user is overwhelmingly likely to end up cancelling the invocation that is trying to download the file (and the subsequent ones won't try at all because they'll see that it's already checked within 24 hours).
3. Even if it's configured to let the previous execution finish instead of cancelling, if the user has typed more characters it will still throw away that output and re-run the workflow, which will have the same effect of not letting the user see the update.

I think a much better approach here is, when it's determined that the interval has passed, to fork a new process that does all the networking, and when it has finished downloading the new release (assuming there is one) it can then post a sticky notification to notification center that the user can select to download it. That said, I haven't actually looked at the APIs for macOS notifications so I'm not sure if a command-line process can actually just post a notification, wait around for a response, and handle the action, or if you really have to be using a .app with a unique bundle identifier for this.

If this won't actually work then instead I'd suggest having the forked process simply check for a new version but not actually download it yet, and if it finds one it can then write the download URL to a special spot in the workflow's cache folder, and then broadcast a regular notification saying there's an update and to trigger the workflow again to see it (I know command-line tools can send notifications, I just don't know if they can handle actions on them). Then the updater module would check for the presence of that file, and if it finds one, it would spit out the XML to download and install the update. When that action is triggered, it would EITHER have to feed back into rust code that downloads the file then opens it, or feed into a bash script that uses something like `curl -O` on the URL to download it and then opens it. Note that the XML offering to install the update would also have to have an option to ignore the update (which would write out the version or URL to ignore back to the cache folder, so subsequent runs of the update check can see that and properly ignore the update).

In any case, the ultimate goal here is 1) never block the workflow output, and 2) make sure the workflow behaves properly if its output is thrown away and the workflow is re-invoked.

spamwax commented:
Thanks for the detailed feedback. I totally agree with you about blocking the workflow and its impact on user experience.

- I agree that the PR can benefit from some of the suggestions here, but my main goal was to just provide a way for your crate's user to check for updates and *if necessary* **and asked for** download them. 
That's why I am just providing basic APIs as I thought since this is a library I should not decide for its users how thread forking and dealing with notifications is done.
- The workflow will not block every 24hrs even if there is an update unless *library* user explicitly asks for the download operation to take place. The example I have added to `updater` module could have been the reason you got that impression and probably it should get updated to reflect your points.
- Current PR can be used in a workflow to add a dedicated Script Filter or Keyword action to initiate a full download and installation, which presumably will be separate than workflow's normal behavior.
- Right now it's up to crate's users how they want to deal with blocking issue by possibly spawning a new thread to deal with download operations. In my own workflow based on this crate, within intervals I talk to github just to see if a newer version is available without downloading. This is done in a separate thread which is then joined by main to get the results. Then a notification is sent to user only if an update is available, which will then require their explicit action to start the download and upgrade process.
- I believe `alfred-rs` is mainly used to build binary applications. Correct me if I am wrong, but Rust by design will kill any thread that's spawned as soon as `main` exits. This will force us to fork from the binary application and run a process as a daemon when download operation is needed. It is probably doable but would add complexity to this crate (and probably its users).
- You are right about scenarios where interruptions happen and I can see it's possible that every keystroke can fire a network request to ask for latest release's metadata. I will update with a fix to this and other issues raised in your comment.

spamwax committed:
Address documentation change requests.

spamwax committed:
Minor improvements based on PR feedback.

- Only use Into<String>
- Move a use statement to top of file
- Ensure a repo name using for test mocking is unique
- Don't talk to network within a test.

spamwax committed:
Replace RefCell with Cell

spamwax committed:
Use Option to replace epoch time checks.

spamwax committed:
Don't persis version onto disk if set during run time.

spamwax committed:
Improved it_loads_last_updater_state() test.

spamwax committed:
Update doc examples.

spamwax committed:
Rename Releaser trait's method to latest_version.

- update_ready() now saves the info related to last release and its
availablitly to a cache file. This will let library user know if there
were update available or not between interval checks.

spamwax committed:
Use url crate.

spamwax committed:
Aggressively sanitize workflow's name used in filenames.

spamwax committed:
Squash some small commits.

impl From<S> for GithubReleaser

Better use of rust features!

Update some docs based on new changes to API

Doc update.

Add test for building file name

Remove failure_derive.

kballard commented:
@spamwax Hi sorry for not responding, I've been super busy lately. I'll try to get back to this soon, though I don't know if I'll have the time until this weekend.

spamwax committed:
Move private methods to its own module.

spamwax committed:
Return Result of set_interval.

- Hide docs for private impls.
- Update tests related to set_interval change.

spamwax committed:
Instead of hiding docs properly expose imp to its parent module only.

spamwax committed:
TODO list update.

kballard commented:
I just converted this PR over to using [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8) because I think it's a lot better than GitHub's built-in PR tools.

---







*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBC_d275UC_2u2wrBFm:b-dt57ce)*
<!-- Sent from Reviewable.io -->

kballard commented:
**Note:** The following comment applies to your original PR. I haven't reviewed your recent changes yet.

---

> I agree that the PR can benefit from some of the suggestions here, but my main goal was to just provide a way for your crate's user to check for updates and if necessary and asked for download them.
That's why I am just providing basic APIs as I thought since this is a library I should not decide for its users how thread forking and dealing with notifications is done.

There's two problems with this approach:

1. If we provide an API that allows users to shoot themselves in the foot, they absolutely will shoot themselves in the foot.
2. The API you provided spits out an alfred XML blob, which means it can only be used in the synchronous case. Users cannot use this with the fork approach I proposed because it simply won't work that way.

> The workflow will not block every 24hrs even if there is an update unless library user explicitly asks for the download operation to take place.

Not sure what you mean. It definitely has to block just to check if there's a new update, even if it doesn't download.

> Right now it's up to crate's users how they want to deal with blocking issue by possibly spawning a new thread to deal with download operations.

At best the user can check for a new update concurrently with any other work they do, but ultimately it still has to block on the network request if the other work it's doing finishes first.

> You are right about scenarios where interruptions happen and I can see it's possible that every keystroke can fire a network request to ask for latest release's metadata. I will update with a fix to this and other issues raised in your comment.

My problem wasn't that every keystroke would fire a network request (although that would be a problem), which it won't because your API wrote the new update timestamp to disk before actually checking for an update. The problem was the user would probably not see the update at all because they'll have typed another character which would either kill the previous instance (that was checking for an update), or wait for it to finish and then throw away its results.

---

I'll review your recent changes now.

---







*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBCaLAB9Dta2_IW1q97:bctzdo)*
<!-- Sent from Reviewable.io -->

kballard commented:
As before, I appreciate what you're trying to do with this PR. I still have reservations about the way it works though, since checking for a new release is still a blocking operation. You have to assume any network activity could take on the order of seconds before getting any kind of response (even an error). Imagine if your user is tethered to a cell phone which itself is on a poor EDGE connection.

I'm also not sure if it actually makes sense to put this functionality in the `alfred` crate directly. Turning on the feature flag doesn't affect any of the existing crate behavior, it just adds a new module. It seems to me this could simply be done as a separate crate entirely which depends on `alfred`. I'm happy to update the README to recommend the use of another crate for implementing workflow update support.

---

Reviewed 2 of 6 files at r1, 2 of 4 files at r2, 3 of 3 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[src/lib.rs, line 13 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_6-r1-13:-LBCa2zPAt3EPy-wOTOG:b-mfz6c2) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/lib.rs#L13)):*
<details><summary><i>Previously, spamwax wrote…</i></summary><blockquote>

I think this is working because in next line I have defined what `[updater]` is pointing to.
However I ran to similar issues when using nightly/stable. It seems rustdoc is moving to a new markdown backend and this is related to the transition.
</blockquote></details>

It looks like GHFM (and CommonMark) defines a new [shortcut reference link](https://github.github.com/gfm/#shortcut-reference-link) which is what you're using here. I haven't seen this before. It must have been newly-defined for CommonMark and then adopted by GHFM.

---

*[src/updater/mod.rs, line 320 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-320:-LBCf06o6YZTDd72NiPs:b-dnhwnl) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L320)):*
<details><summary><i>Previously, spamwax wrote…</i></summary><blockquote>

I included this so workflow authors can let workflow users change the interval. 
</blockquote></details>

Workflow authors that want to let users do that can build their own system for persisting the user's choice to disk. I really think this library shouldn't try to save to disk because it's surprising and the error story isn't very good (e.g. this method returns a `Result`, but setting the interval always works in code, so returning an error is mildly confusing, and there's also no way when initializing the updater for the workflow to tell if the interval was customized). Any workflow that wants to customize this can just do so itself.

---

*[src/updater/mod.rs, line 645 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-645:-LBChnh79x7vLYxibFGf:b4t3f7y) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L645)):*
<details><summary><i>Previously, spamwax wrote…</i></summary><blockquote>

we could, this was here to test actual errors that talking to a real server may cause. As you can see it's intended to be run locally. I can add an `#[ignore]` to it.
</blockquote></details>

That doesn't really solve anything. If you want to test the actual errors, then mock out the network call to return the error you're trying to test the behavior of.

---

*[src/updater/mod.rs, line 71 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCcAMq1TE5kjad9Rwy:-LBCcAMq1TE5kjad9Rwz:b-y1mtz8) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L71)):*
> ```Rust
> //! // Depending on how long has elapsed since first run consequent calls
> //! // to `update_ready()` may return false if it has been less than
> //! // interval set for checking (defaults to 24 hours).
> ```

"interval" should be "the interval"

---

*[src/updater/mod.rs, line 133 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCiGMl0di_Fu45C2UX:-LBCiGMl0di_Fu45C2UY:bja3y0p) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L133)):*
> ```Rust
> pub use self::releaser::Releaser;
> 
> // TODO: get the load_or_new patch from other branch
> ```

I'm just flagging this because this should be removed in the final version.

---

*[src/updater/mod.rs, line 288 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCdqO0B7OuVXI29sBH:-LBCdqO0B7OuVXI29sBI:byxchwb) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L288)):*
> ```Rust
>     /// # env::set_var("alfred_workflow_uid", "abcdef");
>     /// # env::set_var("alfred_workflow_data", env::temp_dir());
>     /// # env::set_var("alfred_workflow_version", "0.0.0");
> ```

With `no_run` you can presumably remove all these `set_var` lines since they were just to make it not error out when running.

---

*[src/updater/mod.rs, line 296 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCd_9vB-N1L5sUfOUk:-LBCd_9vB-N1L5sUfOUl:bdtjvw4) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L296)):*
> ```Rust
>     /// # fn main() {
>     /// #     ex_set_version();
>     /// # }
> ```

With `no_run` specified, do you still need to define a `main()` function?

---

*[src/updater/mod.rs, line 322 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCe5WaCsx86R5PPjvz:-LBCe5WaCsx86R5PPjw-:b-8j6mh) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L322)):*
> ```Rust
>     /// # env::set_var("alfred_workflow_uid", "abcdef");
>     /// # env::set_var("alfred_workflow_data", env::temp_dir());
>     /// # env::set_var("alfred_workflow_version", "0.0.0");
> ```

Same here, with `no_run` these `set_var` lines are presumably unnecessary.

Also in this case, you can remove the `fn main()` as well, and also the `extern crate alfred`. If you don't have any `extern crate` lines, rustdoc automatically inserts one for the current crate, and if you don't define any functions, rustdoc automatically wraps the whole thing in a `fn main() {}`. So this can look like

    /// ```rust,no_run
    /// # use alfred::Updater;
    /// # use std::env;
    /// let mut updater = Updater::gh("spamwax/alfred-pinboard-rs")
    ///                           .expect("cannot initiate Updater");
    /// updater.set_interval(7 * 24 * 60 * 60);
    /// ```

---

*[src/updater/mod.rs, line 375 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCflIkB5aZGBHzlvVh:-LBCflIkB5aZGBHzlvVi:b4bgydb) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L375)):*
> ```Rust
>                 serde_json::to_writer(buf_writer, &version)?;
>                 Ok(())
>             })?;
> ```

If we create the file, but get an error writing to it, we should delete the file. If we get an error on creation, we probably shouldn't try to delete it.

---

*[src/updater/mod.rs, line 393 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCgAWMAg63QCmv6nuC:-LBCgAWMAg63QCmv6nuD:b-j1bkw7) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L393)):*
> ```Rust
>             self.releaser
>                 .latest_version()
>                 .and_then(|v| Ok((*self.current_version() < v, v)))
> ```

This can be a `map`.

---

*[src/updater/mod.rs, line 400 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCgREY98usEbZF9gqn:-LBCgREY98usEbZF9gqo:b-jqvny0) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L400)):*
> ```Rust
>                 .and_then(|r| {
>                     self.set_last_check(Utc::now());
>                     self.save()?;
> ```

Why are we writing the cached version to a separate file than the last update check time? Seems to me we could write them both to the same file and simplify things a bit.

---

*[src/updater/mod.rs, line 419 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCght44HzCRCJCp2_d:-LBCght44HzCRCJCp2_e:b-4viaxr) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L419)):*
> ```Rust
>             // update is ready or not until the next due check is upon us.
>             match read_last_check_status() {
>                 Err(_) => ask_releaser_for_update(),
> ```

If the cached version is corrupted we should just treat it as "no newer version". Otherwise if there's some error that causes it to be corrupted or missing every time then we'll be doing a network request every single time you run the workflow.

---

*[src/updater/releaser.rs, line 15 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_9-r1-15:-LBCj6CE1CsEETETrHUS:b-lat3gl) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/releaser.rs#L15)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

This should probably be something that's not actually a potentially-valid repo.
</blockquote></details>

I was actually thinking just make it something not in `foo/bar` form, such as just replacing the value with a manually-generated UUID, but I suppose this works.

---

*[src/updater/releaser.rs, line 68 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCjUonEQfiAEDG0yq6:-LBCjUonEQfiAEDG0yq7:bqctzao) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/releaser.rs#L68)):*
> ```Rust
> }
> 
> impl<S> From<S> for GithubReleaser
> ```

It seems kind of weird to me to use `From` to convert from a string-like thing into a `GithubReleaser`. Normally it's used to convert between types that are conceptually similar (such as between error types, or from a `String` into a `PathBuf`).

---

*[src/updater/releaser.rs, line 108 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCkCylB6nakiGJ0APn:-LBCkCym4lN38TKLVi42:bvk2ndx) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/releaser.rs#L108)):*
> ```Rust
>                 let mut latest: ReleaseItem = serde_json::from_reader(resp)?;
>                 if latest.tag_name.starts_with('v') {
>                     latest.tag_name = latest.tag_name.as_str()[1..].to_string();
> ```

You could just say `latest.tag_name.remove(0);` instead of allocating a copy of the string.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBCkWiBCQZPvMvsIqTp:b7e9sih)*
<!-- Sent from Reviewable.io -->

spamwax commented:
First off, I accidentally pushed this as it wasn't ready yet since I am working on providing a method that will asynchronously check for update while returning a handle to a mpsc channel.

The constraint that I have is that any workflow, using this crate or not, can only live as long as the user hasn't cancelled the Alfred's window. You could argue that it'd be possible to run a background download daemon when the user is on a poor EDGE connection. My issue with such approach is that we are writing *Alfred* workflows not system services, and shouldn't do anything if user wishes to close Alfred's window.

The new asynchronous method will work like this:

1- Our binary executable is loaded and run by alfred
2- We spawn a new thread using `update_ready_async` to check for an update and return handle to the comm. channel.
3- Our app does its work and produces some `alfred::items`
4- User can now check on thread by using the non-blocking `try_recv`. If it's ok and an update is available, they can then add another `alfred::item` to the current list or invoke a sticky notifications
5- If `try_rcev` wasn't successful, the app exits, and you are right, the user will not be notified of new update.

> The problem was the user would probably not see the update at all because they'll have typed another character which would either kill the previous instance (that was checking for an update), or wait for it to finish and then throw away its results

Do you have a solution for this problem without forking a brand new process in the background that will live even when Alfred's window is closed?

I will push my latest commits (including `update_ready_async`) and fix the doc issues raised here.



 

---







*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBFKc7f6p3nrnDwvDv9:bw617z1)*
<!-- Sent from Reviewable.io -->

spamwax commented:
*[src/updater/releaser.rs, line 108 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCkCylB6nakiGJ0APn:-LBFZHZU0xrDqLHwhvb6:b-896fix) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/releaser.rs#L108)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

You could just say `latest.tag_name.remove(0);` instead of allocating a copy of the string.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8)*
<!-- Sent from Reviewable.io -->

spamwax commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/updater/mod.rs, line 320 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-320:-LBFRuUECGtXgjfsKTqV:b-16lshb) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L320)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

Workflow authors that want to let users do that can build their own system for persisting the user's choice to disk. I really think this library shouldn't try to save to disk because it's surprising and the error story isn't very good (e.g. this method returns a `Result`, but setting the interval always works in code, so returning an error is mildly confusing, and there's also no way when initializing the updater for the workflow to tell if the interval was customized). Any workflow that wants to customize this can just do so itself.
</blockquote></details>

ok, I'll remove this. I was actually going to do another PR that will add a feature to let this library users  store/load workflow's config/setting without rolling their own serializer/deserializer and dealing with disk operation, but I guess this answers that PR :laughing: 

---

*[src/updater/mod.rs, line 416 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-416:-LBFSkHA38J7rLkF8mEu:b-896fix) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L416)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

You should insert "the" before "latest workflow" ("Downloading the latest workflow can fail")
</blockquote></details>

Done.

---

*[src/updater/mod.rs, line 472 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-472:-LBFSsn5ASzZWSBe9ESO:b-896fix) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L472)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

If this fails due to a non-transient error, this will cause the workflow to crash on launch forever (as it will attempt to save again every single time you run it).
</blockquote></details>

Done.

---

*[src/updater/mod.rs, line 645 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-645:-LBFYv55B3nhj46MbzrM:bctex74) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L645)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

That doesn't really solve anything. If you want to test the actual errors, then mock out the network call to return the error you're trying to test the behavior of.
</blockquote></details>

I'll remove this, the mocker I am using doesn't simulate network delays and sudden connection drops.

---

*[src/updater/mod.rs, line 288 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCdqO0B7OuVXI29sBH:-LBFTVAF6wwenwEx0R7b:b-1f8sgq) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L288)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

With `no_run` you can presumably remove all these `set_var` lines since they were just to make it not error out when running.
</blockquote></details>

will remove `no_run`

---

*[src/updater/mod.rs, line 296 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCd_9vB-N1L5sUfOUk:-LBFTHmC07FMP0OzjRuy:b7nu76s) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L296)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

With `no_run` specified, do you still need to define a `main()` function?
</blockquote></details>

my guess is yes as I am using `?` operator. Rust didn't used to allow it in `main()`

---

*[src/updater/mod.rs, line 322 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCe5WaCsx86R5PPjvz:-LBFTe57Azm78EFeihIH:b-m88dyr) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L322)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

Same here, with `no_run` these `set_var` lines are presumably unnecessary.

Also in this case, you can remove the `fn main()` as well, and also the `extern crate alfred`. If you don't have any `extern crate` lines, rustdoc automatically inserts one for the current crate, and if you don't define any functions, rustdoc automatically wraps the whole thing in a `fn main() {}`. So this can look like

    /// ```rust,no_run
    /// # use alfred::Updater;
    /// # use std::env;
    /// let mut updater = Updater::gh("spamwax/alfred-pinboard-rs")
    ///                           .expect("cannot initiate Updater");
    /// updater.set_interval(7 * 24 * 60 * 60);
    /// ```

</blockquote></details>

will remove `no_run`, none of the lines in example does any network/disk activity so run should be ok.

---

*[src/updater/mod.rs, line 400 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCgREY98usEbZF9gqn:-LBFUr3P4wwgf-Bop3HT:b-lojqxo) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L400)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

Why are we writing the cached version to a separate file than the last update check time? Seems to me we could write them both to the same file and simplify things a bit.
</blockquote></details>

just to reduce chance of corrupting our main file which holds version, timestamp and update interval vs this one, which we will try to update every 24hrs.

---

*[src/updater/mod.rs, line 419 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCght44HzCRCJCp2_d:-LBFVuQODjSyJxDhfkAt:bj2dmbb) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L419)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

If the cached version is corrupted we should just treat it as "no newer version". Otherwise if there's some error that causes it to be corrupted or missing every time then we'll be doing a network request every single time you run the workflow.
</blockquote></details>

I'll remove the functionality. The reason I added because I thought one of the major reasons the file will get corrupted is when we successfully received a response from network and were busy caching it when user cancels the workflow.

---

*[src/updater/releaser.rs, line 68 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCjUonEQfiAEDG0yq6:-LBFWZOJ38-2ChCIpYpz:b-yzt2n9) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/releaser.rs#L68)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

It seems kind of weird to me to use `From` to convert from a string-like thing into a `GithubReleaser`. Normally it's used to convert between types that are conceptually similar (such as between error types, or from a `String` into a `PathBuf`).
</blockquote></details>

I ran to a visibility issue of function named `new`. The name was shared between releaser and private method of GithubReleaser.
I'll remove this.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8)*
<!-- Sent from Reviewable.io -->

spamwax commented:
Review status: all files reviewed at latest revision, 17 unresolved discussions.

---

*[src/updater/mod.rs, line 179 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-179:-LBFZXBPB_uPOOkFd2gh:b-896fix) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L179)):*
<details><summary><i>Previously, kballard (Kevin Ballard) wrote…</i></summary><blockquote>

You should probably insert "The" before "Updater"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8)*
<!-- Sent from Reviewable.io -->

kballard commented:
> You could argue that it'd be possible to run a background download daemon when the user is on a poor EDGE connection. My issue with such approach is that we are writing Alfred workflows not system services, and shouldn't do anything if user wishes to close Alfred's window.

This isn't just about EDGE, that was simply to illustrate the worst case. This is also just the fact that for most Alfred workflows, blocking until a network request has finished is a problem. Your update check here is designed to operate at the script filter level (rather than running when the user has triggered the final action), so it needs to be really quick. You could theoretically design it such that you're supposed to check for update when the user actually triggers an action, just to cache the results, and then strictly check the cache when running the script filter, but this has the downside that workflow that is used less often than it's updated will only ever report old versions to the user (because by the time they see the cached update result, it's already stale).

> Do you have a solution for this problem without forking a brand new process in the background that will live even when Alfred's window is closed?

No.

I don't understand your objection to forking a process. It's not a big deal. It's not like we're starting up a long-running background process, it's just a short-lived check.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[src/updater/mod.rs, line 400 at r3](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBCgREY98usEbZF9gqn:-LBr1mZ03dfL0Fzew-Bs:bl9j48n) ([raw file](https://github.com/kballard/alfred-rs/blob/7a2591e85a64cab1ae18c47c3b64d6eb54f5014b/src/updater/mod.rs#L400)):*
<details><summary><i>Previously, spamwax wrote…</i></summary><blockquote>

just to reduce chance of corrupting our main file which holds version, timestamp and update interval vs this one, which we will try to update every 24hrs.
</blockquote></details>

If you're worried about corrupting the file, you should write to a temporary, then atomically swap it into place.

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBr0_gh2SYuIRSQYMPZ:bvqvfcm)*
<!-- Sent from Reviewable.io -->

spamwax commented:
Kevin, thanks for your feedback. I am learning from your comments and our interaction here and that is appreciated.

I have two main concerns related to forking:
- We are operating in Alfred world, if user closes its window we should honor that and stop all operations. Of course it's a different story if there is an action that user explicitly asks for a long download process.
- In a quick search that I did, it seems Rust does not directly support forking yet, if you have any pointer on how to do this in an idiomatic way please let me know.

Additionally:
- This PR can be used for both Script Filter **and** other Alfred action items. Yes it has the potential of momentarily blocking the workflow (as you mentioned *short-lived*) if it is used in Script Filter as does any other disk IO operation from any workflow that may happen during a heavy disk operation on OS side.
- Based on looking at other Alfred libraries [written in Go](https://github.com/deanishe/awgo) and Python, it seems people are generally settled on the convention of not forking out.
- As I mentioned before, my new upcoming additions will let user start a thread to check for update. If the thread is done by the time workflow binary is about to exit then user can take advantage of results it would return. If it is not done, no blocking will happen and no update will be reported.
- If it is just a short-lived check we can get away with it every 24 hrs based on above thread-based solution. Other workflows such as [TLDR](https://github.com/cs1707/tldr-alfred) and [Google Search](http://zhaocai.github.io/alfred2-google-workflow/) do have these quick and short hiccups as user is typing.

And the end this is your crate 😄 so if you don't think this functionality is suited for crate's purpose please let me know. As you suggested I can create another crate that depends on this one and adds the functionality.

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBsZg9ZAkVGQb5pfIrB:b-9jcr3l)*
<!-- Sent from Reviewable.io -->

spamwax commented:
BTW as you said in one your comments: "if a workflow author wants to let the user 

---

Review status: all files reviewed at latest revision, 15 unresolved discussions.

---

*[src/updater/mod.rs, line 320 at r1](https://reviewable.io/reviews/kballard/alfred-rs/8#-LBC_SF5pLrZpcUvoL_8-r1-320:-LBshZUx97O5qqUKPrk0:bu0i4j2) ([raw file](https://github.com/kballard/alfred-rs/blob/cc3c24b139a95f03aff5947cb77900ee32d7c57d/src/updater/mod.rs#L320)):*
> Workflow authors that want to let users do that can build their own system for persisting the user's choice to disk.

Can we also apply the same logic and say authors should build their own system for forking and not blocking, this crate only wraps network and version comparison functionalities and provides them in an 'updater' package?

> I really think this library shouldn't try to save to disk because it's surprising and the error story isn't very good.

Thinking more about this, I think you can apply that to entire of this PR as network operations are more surprising!

---


*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBsh7RfEd3MrPctcTNl:b-f07qyh)*
<!-- Sent from Reviewable.io -->

kballard commented:
> We are operating in Alfred world, if user closes its window we should honor that and stop all operations. Of course it's a different story if there is an action that user explicitly asks for a long download process.

The problem with this is there's no way to know when they closed the window. And using a single execution of a script filter isn't good enough, because that tends to run many times before the window is closed (and blocking that makes for a worse user experience overall).

That said, checking for a new version should be sufficiently fast that, despite spawning a background process in order to do so, that process will likely finish its work and exit before the user has closed the Alfred window.

I agree that it shouldn't *download* the update in that process without the user requesting. But merely checking for an update is a) benign, and b) should normally be fast. And you can probably even set a timeout on the network request to ensure that if, say, 10 seconds elapses without getting a result, it considers that a failure and aborts.

> In a quick search that I did, it seems Rust does not directly support forking yet

There may not be a `libstd` method for it, but there's always the [`libc` method](https://docs.rs/libc/0.2.40/libc/fn.fork.html).

That said, I'm not sure if there's any potential issues with the Rust runtime in calling stuff from a forked process without execing. On macOS at least, forking a process only copies the current thread into the new process, so if you call `fork` from a multithreaded program, there's only a small handful of so-called "fork-safe" POSIX APIs you can invoke (basically, the set of functions that are guaranteed to not take a lock, including not allocating). In a single-threaded program there isn't really any worry because without any other threads, you don't run the risk that some lock is being held. However, there is a `pthread_atfork` API that can be used to run code when forking, so again, it's possible the Rust runtime does stuff when forking that means the new process isn't really safe to do anything but exec, though I kind of doubt it.

Given that, there's two potential strategies here:

1. Check if the Rust runtime has any problems with running arbitrary code after a fork. Assuming the answer is "no" (and I expect it is), then document that the call to check for updates must happen before any threads are spawned, and there's probably some way to actually assert this at runtime too.
2. Alternatively, after forking, just immediately re-exec the exact same binary again, but with a special command-line flag or environment variable set to indicate that this is the background process. This of course requires you to have an API to check for the flag/variable in order to change how the program behaves (so that you can hand control of the worker process over to the library in order to do the update check). My recommendation here would be to declare a function that is supposed to be called from `main()` which takes a `FnOnce` that is invoked for the main process and skipped in the background process. This way the workflow's `main()` looks something like
   
   ```rust
   fn main() {
       alfred::updater::set_version("1.23"); // or just read it from workflow variables
       alfred::updater::main(|| {
           // if this executes, this is the main process
       });
   }
   ```

Of course, there's always the third strategy of requiring the workflow to bundle a separate binary that implements the version check, but that requires more setup on the workflow author's part (in order to have two separate binaries). It's a good fallback though if #1 doesn't work and #2 ends up being too awkward.

> Other workflows such as TLDR and Google Search do have these quick and short hiccups as user is typing.

And I consider that to be a terrible UX. I want a solution with zero hiccups.

> And the end this is your crate 😄 so if you don't think this functionality is suited for crate's purpose please let me know. As you suggested I can create another crate that depends on this one and adds the functionality.

On the one hand, it's convenient to have one crate for everything, but on the other hand, it's not very flexible. At this point I think that even if I were to write this functionality myself, I'd put it in a separate crate entirely anyway. So you should probably go that route. Like I said, I'm happy to add a link to your crate in my README.

---







*Comments from [Reviewable](https://reviewable.io/reviews/kballard/alfred-rs/8#-:-LBsiT3mBXfmQRHsgAVT:bybk1bf)*
<!-- Sent from Reviewable.io -->

kballard commented:
I'm going to go ahead and close this PR. I notice you published [spamwax/alfred-workflow](https://github.com/spamwax/alfred-workflow) for this functionality.
